File: BinaryTree/js/levelAverages.js
const levelAverages=(root)=>{
  const levels = []
  fillLevels(root, levels, 0)

  const avg = []

  for(let level of levels){
    avg.push(avg(level))
  }

  return avg
}


const fillLevels =(root, levels, levelNum)=>{
    if(root === null) return

    if(levels.length === levelNum){
        levels[levelNum] = [root.val]
    }else{
        levels[levelNum].push(root.val)
    }

    fillLevels(root.left, levels, levelNum+1)
    fillLevels(root.right, levels, levelNum+1)
}

const avg =(array)=>{
    let sum = 0 
    for(let num of array){
        sum += num
    }

    return sum / array.length
}

File: BinaryTree/go/levelAverages.go
package binarytree

// LevelAverages returns the average value for each level in the tree.
func LevelAverages(root *TreeNode) []float64 {
	if root == nil {
		return []float64{}
	}

	averages := make([]float64, 0, 8)
	queue := []*TreeNode{root}

	for len(queue) > 0 {
		levelSize := len(queue)
		sum := 0

		for i := 0; i < levelSize; i++ {
			node := queue[0]
			queue = queue[1:]
			sum += node.Val

			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		averages = append(averages, float64(sum)/float64(levelSize))
	}

	return averages
}
